### ISSUE TITLE
Getting frequent TaskCanceledExceptions from simple Fluent API calls

### ISSUE THREAD

**Issue by JadedATB** [Azure/azure-libraries-for-net#395]
We're having significant trouble creating a stable application.  We're getting random/spurious TaskCanceledExceptions from seemingly simple calls:

            ISubscription sub = azure.GetCurrentSubscription();

            return await this.Azure.NetworkInterfaces.GetByResourceGroupAsync(resourceGroup, name);

            return await this.Azure.VirtualMachines.GetByResourceGroupAsync(resourceGroup, name);

            return await this.Azure.ResourceGroups.GetByNameAsync(name);

Sample stack trace:
[2018-07-09 13:38:14] System.Threading.Tasks.TaskCanceledException: A task was canceled.
[2018-07-09 13:38:14]    at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
[2018-07-09 13:38:14]    at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
[2018-07-09 13:38:14]    at Microsoft.Azure.Management.ResourceManager.Fluent.SubscriptionsOperations.<GetWithHttpMessagesAsync>d__6.MoveNext()
[2018-07-09 13:38:14] --- End of stack trace from previous location where exception was thrown ---
[2018-07-09 13:38:14]    at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
[2018-07-09 13:38:14]    at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
[2018-07-09 13:38:14]    at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
[2018-07-09 13:38:14]    at Microsoft.Azure.Management.ResourceManager.Fluent.SubscriptionsOperationsExtensions.<GetAsync>d__1.MoveNext()
[2018-07-09 13:38:14] --- End of stack trace from previous location where exception was thrown ---
[2018-07-09 13:38:14]    at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
[2018-07-09 13:38:14]    at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
[2018-07-09 13:38:14]    at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
[2018-07-09 13:38:14]    at Microsoft.Azure.Management.ResourceManager.Fluent.Core.Extensions.Synchronize[TResult](Func`1 function)
[2018-07-09 13:38:14]    at Microsoft.Azure.Management.ResourceManager.Fluent.SubscriptionsImpl.GetById(String id)

TaskCanceledExceptions broke 9 of our deployment tests in our nightly build last night.  All the same tests ran fine the night before.

More full stack traces can be sent privately if requested.

**Comment by hovsepm** [comment]
@JadedATB could you provide some more context on your scenario. 

Azure Cloud have a limit on number of requests per minute that any client app can do and if your application hits that by doing too many parallel polling requests to determine the state of the resources during deployment you may be simply throttled by Azure.

**Comment by JadedATB** [comment]
We're building a tool to edit/deploy specific types of Azure environments,
currently supporting about 25 Azure resources.  We're deploying most via
the Fluent API, but also using the REST APIs where the Fluent API doesn't
yet exist (e.g. Azure Data Factories using
Microsoft.Azure.Management.DataFactory NuGet package).

We have a nightly build that runs deployment tests for all resource types
(41 tests in all).  The tests are generally straightforward.  If my
information is up-to-date the request limits are something like 15,000
reads and 1,200 writes per hour, correct?  I would be shocked if we had
anywhere near that many requests.  Is there a way to check that?  I thought
those were also generally identifiable by 429 error codes.  And if it were
simply a throttling issue, I would expect all my tests to fail after a
certain point, but it was 9 out of 41 (non-sequential).

**Comment by hovsepm** [comment]
What is your polling interval for each resource? Does it happen in parallel?

**Comment by JadedATB** [comment]
What do you mean by polling interval?  The tests are not currently running in parallel.  They run for about 5 hours every night.  Most of that time is waiting for certain resources to deploy (API Management test takes around 45 minutes, Redis cache takes about 30 minutes, app gateway takes about 30 minutes, etc.)  We also have tests that deploy some long-running script extensions on virtual machines.

Our tests are set up to deploy a resource, make sure we don't get any exceptions from the deployment, then deprovision it.

**Comment by hovsepm** [comment]
@JadedATB please contact me via IM. I'm still not clear about what is your scenario.

**Comment by hovsepm** [comment]
@JadedATB or could you share your test sources?

**Comment by hovsepm** [comment]
Closing as non-repro. Please open a new issue if you hit this in v1.16.1 or higher.

**Comment by TimLovellSmith** [comment]
For everyone's info... We also hit TaskCanceledException all the time in Fluent SDK. It seemed to be because its HttpClient stack uses task cancellations to use timeouts - so I wrote a wrapper function for all our Fluent SDK calls so I never needed to deal with ti again.... But actually the upshot of that is that we still have to manually fix up our code every time we *forget* to use the wrapper function. Should have sent a PR to addressed the root problem huh? #Hindsight=20/20 But on the upside, at least our wrapper functional also knows how to deal with some other throttling issues by auto-retry and exponential backoff...
